# K-point and cutoff convergence for Li bcc surfaces

This notebook tests plane-wave cutoff (`ecut`) and k-point mesh convergence for
Li adatom on bcc Li surfaces: (110), (111), and (100).

It assumes you already have relaxed **initial** adatom states for each surface
saved as trajectory files (e.g. from your NEB pre-relaxation). You can adjust
the file paths in the configuration cell below.

In [1]:

import os
import numpy as np
import pandas as pd

from ase.io import read
from gpaw import GPAW, PW

# === User configuration ===
# Adjust these to match your project layout / choices.

# Exchange–correlation functional (set to whatever you used in your NEB)
xc = "PBE"  # e.g. "PBE", "LDA", or a GPAW xc string

# Directory to write GPAW output text files
output_dir = "crystal_structure_neb_kconv"
os.makedirs(output_dir, exist_ok=True)

# Paths to relaxed initial adatom states for each surface.
# TODO: update these paths to your actual files.
state_files = {
    "Na": "multi_neb_results_restartable_time/Na-initial.traj",
}

# Load Atoms objects
states = {name: read(path) for name, path in state_files.items()}
states


{'Na': Atoms(symbols='Na73', pbc=[True, True, False], cell=[17.94637010651458, 12.690000000000001, 28.97318505325729], tags=..., constraint=FixAtoms(indices=[0, 1, 2, 3, 8, 9, 10, 11, 16, 17, 18, 19, 24, 25, 26, 27, 32, 33, 34, 35, 40, 41, 42, 43, 48, 49, 50, 51, 56, 57, 58, 59, 64, 65, 66, 67]), calculator=SinglePointCalculator(...))}

In [2]:

def sp_energy_with_settings(
    atoms,
    label: str,
    pw_cutoff: float,
    k: int,
    txt_prefix: str = "kconv",
) -> float:
    """Single-point energy with given PW cutoff and kpts=(k, k, 1)."""
    atoms = atoms.copy()  # avoid modifying original
    calc = GPAW(
        mode=PW(pw_cutoff),
        xc=xc,
        kpts=(k, k, 1),
        convergence={"eigenstates": 5e-6, "density": 1e-4},
        txt=os.path.join(
            output_dir,
            f"{txt_prefix}-{label}-ecut{int(pw_cutoff)}-k{k}.txt",
        ),
    )
    atoms.calc = calc
    E = atoms.get_potential_energy()
    return E


In [ ]:

# === Convergence grid ===
pw_list = [300, 400, 500, 600]  # eV
k_list  = [2, 3, 4]             # (2x2x1), (3x3x1), (4x4x1)

results = []

for surf_label, atoms in states.items():
    print(f"=== Surface: {surf_label} ===")
    for pw in pw_list:
        for k in k_list:
            E = sp_energy_with_settings(
                atoms,
                label=surf_label,
                pw_cutoff=pw,
                k=k,
                txt_prefix="kconv",
            )
            results.append(
                {
                    "surface": surf_label,
                    "ecut_eV": pw,
                    "kpts": f"{k}x{k}x1",
                    "k": k,
                    "E_eV": E,
                }
            )
            print(f"  ecut={pw:4d} eV, k={k}x{k}x1 -> E = {E:.6f} eV")

df_conv = pd.DataFrame(results)
df_conv


=== Surface: Na ===
  ecut= 300 eV, k=2x2x1 -> E = -88.713074 eV
  ecut= 300 eV, k=2x2x1 -> E = -88.713074 eV
  ecut= 300 eV, k=3x3x1 -> E = -88.943402 eV
  ecut= 300 eV, k=3x3x1 -> E = -88.943402 eV
  ecut= 300 eV, k=4x4x1 -> E = -88.933869 eV
  ecut= 300 eV, k=4x4x1 -> E = -88.933869 eV
  ecut= 400 eV, k=2x2x1 -> E = -90.018516 eV
  ecut= 400 eV, k=2x2x1 -> E = -90.018516 eV
  ecut= 400 eV, k=3x3x1 -> E = -90.252323 eV


In [ ]:

def analyse_convergence(df: pd.DataFrame, ref_pw: float = 600, ref_k: int = 4) -> pd.DataFrame:
    """Add column with ΔE (meV) vs reference (ref_pw, ref_k) for each surface."""
    rows = []
    for surf in sorted(df["surface"].unique()):
        df_s = df[df["surface"] == surf].copy()
        ref_row = df_s[(df_s["ecut_eV"] == ref_pw) & (df_s["k"] == ref_k)]
        if len(ref_row) != 1:
            print(f"WARNING: reference (ecut={ref_pw}, k={ref_k}) not found for {surf}")
            continue
        E_ref = ref_row["E_eV"].values[0]
        df_s["dE_meV_vs_ref"] = (df_s["E_eV"] - E_ref) * 1000.0
        rows.append(df_s)
    if not rows:
        raise RuntimeError("No reference rows found; adjust ref_pw / ref_k to match your grid.")
    return pd.concat(rows, ignore_index=True)


df_conv_with_delta = analyse_convergence(df_conv, ref_pw=600, ref_k=4)
df_conv_with_delta.sort_values(["surface", "ecut_eV", "k"], inplace=True)
df_conv_with_delta


surface  ecut_eV   kpts  k        E_eV  dE_meV_vs_ref
0    li100      300  2x2x1  2 -131.405546     422.812264
1    li100      300  3x3x1  3 -131.740208      88.149978
2    li100      300  4x4x1  4 -131.812406      15.951817
3    li100      400  2x2x1  2 -131.412825     415.533004
4    li100      400  3x3x1  3 -131.747458      80.900470
5    li100      400  4x4x1  4 -131.819632       8.725929
6    li100      500  2x2x1  2 -131.420321     408.037472
7    li100      500  3x3x1  3 -131.754989      73.368632
8    li100      500  4x4x1  4 -131.827183       1.174821
9    li100      600  2x2x1  2 -131.421458     406.899975
10   li100      600  3x3x1  3 -131.756190      72.168005
11   li100      600  4x4x1  4 -131.828358       0.000000
12   li110      300  2x2x1  2 -128.642857     470.828506
13   li110      300  3x3x1  3 -129.051134      62.551547
14   li110      300  4x4x1  4 -129.097431      16.253759
15   li110      400  2x2x1  2 -128.650326     463.359545
16   li110      400  3x3x1  3 -129.058228      55.457559
17   li110      400  4x4x1  4 -129.104872       8.813456
18   li110      500  2x2x1  2 -128.657966     455.719140
19   li110      500  3x3x1  3 -129.066075      47.609882
20   li110      500  4x4x1  4 -129.112494       1.191404
21   li110      600  2x2x1  2 -128.659149     454.536317
22   li110      600  3x3x1  3 -129.067331      46.354370
23   li110      600  4x4x1  4 -129.113685       0.000000
24   li111      300  2x2x1  2 -125.212861     -89.563069
25   li111      300  3x3x1  3 -125.179991     -56.693897
26   li111      300  4x4x1  4 -125.108001      15.296893
27   li111      400  2x2x1  2 -125.219786     -96.488446
28   li111      400  3x3x1  3 -125.186895     -63.597851
29   li111      400  4x4x1  4 -125.114935       8.362729
30   li111      500  2x2x1  2 -125.226986    -103.688033
31   li111      500  3x3x1  3 -125.194037     -70.739264
32   li111      500  4x4x1  4 -125.122134       1.163066
33   li111      600  2x2x1  2 -125.228146    -104.848099
34   li111      600  3x3x1  3 -125.195191     -71.893768
35   li111      600  4x4x1  4 -125.123298       0.000000

In [ ]:

# Show a compact table of ΔE (meV) vs (ecut, k) for each surface
from IPython.display import display

for surf in ["li110", "li111", "li100"]:
    print(f"\n=== {surf} ===")
    df_surf = df_conv_with_delta[df_conv_with_delta["surface"] == surf]
    if df_surf.empty:
        print("  (no data for this surface; check state_files / convergence grid)")
        continue
    display(
        df_surf
        .pivot(index="ecut_eV", columns="k", values="dE_meV_vs_ref")
        .rename_axis("ecut_eV (PW)", axis=0)
        .rename_axis("k (k x k x 1)", axis=1)
        .round(3)
    )



=== li110 ===


k (k x k x 1)        2       3       4
ecut_eV (PW)                          
300            470.829  62.552  16.254
400            463.360  55.458   8.813
500            455.719  47.610   1.191
600            454.536  46.354   0.000


=== li111 ===


k (k x k x 1)        2       3       4
ecut_eV (PW)                          
300            -89.563 -56.694  15.297
400            -96.488 -63.598   8.363
500           -103.688 -70.739   1.163
600           -104.848 -71.894   0.000


=== li100 ===


k (k x k x 1)        2       3       4
ecut_eV (PW)                          
300            422.812  88.150  15.952
400            415.533  80.900   8.726
500            408.037  73.369   1.175
600            406.900  72.168   0.000

# Report: K-point and Plane-Wave Cutoff Convergence Study

## Overview

This notebook performs a systematic convergence study for density functional theory (DFT) calculations on Li adatoms adsorbed on bcc Li surfaces. The study tests two critical computational parameters:

1. **Plane-wave cutoff energy (`ecut`)**: Controls the size of the basis set for representing wavefunctions
2. **K-point mesh density (`k`)**: Controls the sampling of the Brillouin zone

Both parameters directly affect the accuracy and computational cost of DFT calculations. Finding the optimal balance between accuracy and computational efficiency is essential for production calculations.

---

## Methodology

### Computational Setup
- **DFT Code**: GPAW (Grid-based Projector Augmented Wave)
- **Exchange-Correlation Functional**: PBE (Perdew-Burke-Ernzerhof)
- **Basis Set**: Plane-wave (PW) mode
- **Convergence Criteria**: 
  - Eigenstates: 5×10⁻⁶
  - Density: 1×10⁻⁴

### Parameter Grid
| Parameter | Values Tested |
|-----------|---------------|
| Plane-wave cutoff | 300, 400, 500, 600 eV |
| K-point mesh | 2×2×1, 3×3×1, 4×4×1 |

### Surfaces Studied
- **Li(110)**: Most stable bcc surface
- **Li(111)**: Higher energy surface
- **Li(100)**: Intermediate stability

### Analysis Method
Energy differences (ΔE) are computed relative to the most converged calculation (ecut = 600 eV, k = 4×4×1) for each surface. Values are reported in meV for easier interpretation.

---

## Results Summary

### Li(110) Surface
| ecut (eV) | k=2 | k=3 | k=4 |
|-----------|-----|-----|-----|
| 300 | 470.8 | 62.6 | 16.3 |
| 400 | 463.4 | 55.5 | 8.8 |
| 500 | 455.7 | 47.6 | 1.2 |
| 600 | 454.5 | 46.4 | 0.0 |

### Li(111) Surface
| ecut (eV) | k=2 | k=3 | k=4 |
|-----------|-----|-----|-----|
| 300 | -89.6 | -56.7 | 15.3 |
| 400 | -96.5 | -63.6 | 8.4 |
| 500 | -103.7 | -70.7 | 1.2 |
| 600 | -104.8 | -71.9 | 0.0 |

### Li(100) Surface
| ecut (eV) | k=2 | k=3 | k=4 |
|-----------|-----|-----|-----|
| 300 | 422.8 | 88.2 | 16.0 |
| 400 | 415.5 | 80.9 | 8.7 |
| 500 | 408.0 | 73.4 | 1.2 |
| 600 | 406.9 | 72.2 | 0.0 |

---

## Key Findings

### 1. K-point Convergence is Critical
The most significant observation is the **dramatic dependence on k-point sampling**:
- Using k=2×2×1 introduces errors of **400-470 meV** for Li(110) and Li(100)
- Even k=3×3×1 shows errors of **46-88 meV** depending on the surface
- Only k=4×4×1 achieves sub-20 meV accuracy across all cutoff energies

**Interpretation**: Metal surfaces require dense k-point sampling due to the delocalized nature of metallic bonding. The 2×2×1 mesh is insufficient for capturing Fermi surface features.

### 2. Plane-Wave Cutoff Shows Systematic Convergence
For k=4×4×1:
- **300 eV**: ~15-16 meV error
- **400 eV**: ~8-9 meV error  
- **500 eV**: ~1.2 meV error
- **600 eV**: Reference (0 meV)

The energy converges monotonically with increasing cutoff, which is expected behavior for plane-wave calculations.

### 3. Surface-Dependent Behavior
- **Li(110) and Li(100)**: Show similar positive ΔE values, indicating the reference calculation has lower energy
- **Li(111)**: Shows **negative** ΔE for low k-points, suggesting qualitatively different electronic structure sensitivity to k-point sampling

### 4. Error Decoupling
The errors from ecut and k-points are approximately **additive**. For example, at (300 eV, k=2):
- K-point error alone (at 600 eV): ~455 meV
- Ecut error alone (at k=4): ~16 meV
- Combined: ~471 meV ≈ sum of individual errors

---

## Recommendations

### For Production NEB Calculations

Based on these results, I recommend:

| Target Accuracy | ecut | k-points | Expected Error |
|----------------|------|----------|----------------|
| **High accuracy** (< 2 meV) | 500+ eV | 4×4×1 | < 1.5 meV |
| **Standard** (< 10 meV) | 400 eV | 4×4×1 | ~9 meV |
| **Quick screening** | 400 eV | 3×3×1 | ~60-80 meV |

### Caveats
1. These convergence parameters are **system-specific**. Different adatom species (K, Na) may require re-convergence testing.
2. NEB barrier heights may be more/less sensitive than total energies—difference testing on barriers is recommended.
3. The 1 meV/atom benchmark is often used for energy differences, but barrier heights may tolerate slightly larger absolute errors if relative trends are captured.

---

## Computational Cost Considerations

Increasing parameters affects cost as:
- **K-points**: Cost scales linearly with number of k-points (k² for 2D sampling)
- **Plane-wave cutoff**: Cost scales as $E_{\text{cut}}^{3/2}$ due to larger basis set

The recommended settings of **ecut=400-500 eV** and **k=4×4×1** represent a good balance between accuracy and computational feasibility for Li surface calculations.